core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/FastDoubleParser.kt

```diff
@@ -0,0 +1,233 @@
+                if (char.isWhitespace()) add(' ')
+
+                // add additional chars if needed
+                for (additional in additionals) {
+                    val lowercase = additional.lowercaseChar()
+                    if (lowercase !in localChars && !localStrings.any { lowercase in it }) {
+                        add(lowercase)
+                    }
+                }
+            }
+
+        /**
+         * Builds a set with the specified string from [localDecimalFormatSymbols] and
+         * its fallback string from [fallbackDecimalFormatSymbols] if it's safe to do so.
+         * [additionals] will be added to the set too, when they're safe to add.
+         */
+        fun ((DecimalFormatSymbols) -> String).fromLocalWithFallBack(vararg additionals: String): Set<String> =
+            buildSet {
+                val getString = this@fromLocalWithFallBack
+                val string = getString(localDecimalFormatSymbols).lowercase()
+                add(string)
+
+                // add fallback string if it's safe to do so
+                val fallbackString = getString(fallbackDecimalFormatSymbols).lowercase()
+                if (!fallbackString.any { it in localChars } && fallbackString !in localStrings) {
+                    add(fallbackString)
+                }
+
+                // Fixes NBSP and other whitespace characters not being recognized if the user writes space instead.
+                if (string.isBlank()) add(" ")
+
+                // add additional strings if needed
+                for (additional in additionals) {
+                    val lowercase = additional.lowercase()
+                    if (!lowercase.any { it in localChars } && lowercase !in localStrings) {
+                        add(lowercase)
+                    }
+                }
+            }
+
+        return NumberFormatSymbols.fromDecimalFormatSymbols(localDecimalFormatSymbols)
+            .withPlusSign(setOf('+'))
+            .withDecimalSeparator(DecimalFormatSymbols::getDecimalSeparator.fromLocalWithFallBack())
+            .withGroupingSeparator(DecimalFormatSymbols::getGroupingSeparator.fromLocalWithFallBack())
+            .withExponentSeparator(DecimalFormatSymbols::getExponentSeparator.fromLocalWithFallBack())
+            .withMinusSign(DecimalFormatSymbols::getMinusSign.fromLocalWithFallBack())
+            .withInfinity(DecimalFormatSymbols::getInfinity.fromLocalWithFallBack(*INFINITIES))
+            .withNaN(DecimalFormatSymbols::getNaN.fromLocalWithFallBack(*NANS))
+    }
+
+    /** Fallback method for parsing doubles. */
+    private fun String.parseToDoubleOrNullFallback(): Double? =
+        when (lowercase()) {
+            in INFINITIES, in PLUS_INFINITIES -> Double.POSITIVE_INFINITY
+
+            in MINUS_INFINITIES -> Double.NEGATIVE_INFINITY
+
+            in NANS -> Double.NaN
+
+            else -> {
+                // not thread safe; must be created here
```

Comment:
why is thread safety important here? Or this is a reason to create an Instance of Number Format?

Reply:
Deephaven holds a single instance of this FastDoubleParser class but can parse columns in parallel. Previously I put `numberFormat` outside the function, but this caused bugs, so I put it inside here and it works fine :). I put the warning here because you might assume you could save an object creation by putting `val numberFormat = NumberFormat.getInstance(locale)` inside the class and reusing it, but no, you cannot.